h2o-core/src/main/java/hex/grid/FilteredHyperSpaceIterator.java

```diff
@@ -0,0 +1,80 @@
+package hex.grid;
+
+import hex.Model;
+import hex.grid.filter.PermutationFilter;
+
+public class FilteredHyperSpaceIterator<MP extends Model.Parameters> implements HyperSpaceWalker.HyperSpaceIterator<MP> {
+  private HyperSpaceWalker.HyperSpaceIterator<MP> _hyperSpaceIterator;
+
+  private PermutationFilter<MP> _permutationFilter;
+
+  /**
+   * Total number of visited permutations, including those filtered out by {@code _permutationFilter}.
+   */
+  private long _numberOfVisitedPermutations = 0;
+
+  /**
+   * Keeps number of returned to the user permutations as not all visited permutations are considered to be worthy
+   * for evaluation due to a {@code _permutationFilter}. 
+   */
+  private long _numberOfUsedPermutations = 0;
+  private final long _maxHyperSpaceSize;
+
+  public FilteredHyperSpaceIterator(HyperSpaceWalker.HyperSpaceIterator<MP> hyperSpaceIterator, PermutationFilter<MP> permutationFilter, long maxHyperSpaceSize) {
+    _hyperSpaceIterator = hyperSpaceIterator;
+    _permutationFilter = permutationFilter;
+    _maxHyperSpaceSize = maxHyperSpaceSize;
+  }
+
+  @Override
+  public MP nextModelParameters(Model previousModel) {
+    MP nextMPs = _hyperSpaceIterator.nextModelParameters(previousModel);
+    _numberOfVisitedPermutations++;
+    if(! _permutationFilter.test(nextMPs)) {
+      return hasNext(previousModel) ? nextModelParameters(previousModel) : null;
```

Comment:
with a very large hyperspace and tight filter this may end up with stack-overflow, would prefer using a cycle

Reply:
good point

Reply:
Java 8 sucks. It does not have tail recursion optimisation.

Reply:
@honzasterba fixed